File: src/cli/tree.test.ts
import { test } from "node:test";

import { expect } from "expect";

import {
  type Coordinates,
  InMemoryPackageSource,
  resolveTransitive,
  toCoordinates,
} from "../packages/index.ts";
import { buildForest, formatTree, type TreeSection } from "./tree.ts";

const coord = (spec: string): Coordinates => {
  const [g, a, v] = spec.split(":");
  return toCoordinates(g!, a!, v!);
};

function source(): InMemoryPackageSource {
  const pkg = (spec: string, deps: string[] = []) => ({
    coordinates: coord(spec),
    dependencies: deps.map(coord),
  });
  return new InMemoryPackageSource("registry", [
    pkg("org.x:app:1.0", ["org.x:lib:2.0", "org.y:util:3.0"]),
    pkg("org.x:lib:2.0", ["org.y:util:3.0"]),
    pkg("org.y:util:3.0"),
  ]);
}

test("buildForest nests transitive dependencies under their requester", async () => {
  const resolution = await resolveTransitive([coord("org.x:app:1.0")], [source()]);
  // org.y:util:3.0 is reached first via org.x:app (nearest wins), so it nests
  // there and not again under org.x:lib.
  expect(buildForest(resolution)).toEqual([
    {
      coordinate: "org.x:app:1.0",
      dependencies: [
        { coordinate: "org.x:lib:2.0", dependencies: [] },
        { coordinate: "org.y:util:3.0", dependencies: [] },
      ],
    },
  ]);
});

test("buildForest marks declared roots that no source can resolve", async () => {
  const resolution = await resolveTransitive([coord("org.z:missing:9.9")], [source()]);
  expect(buildForest(resolution)).toEqual([
    { coordinate: "org.z:missing:9.9", dependencies: [], unresolved: true },
  ]);
});

test("formatTree renders one indented section per non-empty configuration", () => {
  const sections: TreeSection[] = [
    {
      configuration: "api",
      tree: [
        {
          coordinate: "org.x:app:1.0",
          dependencies: [
            { coordinate: "org.x:lib:2.0", dependencies: [] },
            { coordinate: "org.y:util:3.0", dependencies: [], unresolved: true },
          ],
        },
      ],
    },
    { configuration: "implementation", tree: [] }, // empty: skipped
  ];
  expect(formatTree(sections)).toBe(
    [
      "api",
      "└── org.x:app:1.0",
      "    ├── org.x:lib:2.0",
      "    └── org.y:util:3.0 (unresolved)",
      "",
    ].join("\n"),
  );
});

test("formatTree reports when nothing is declared", () => {
  expect(formatTree([{ configuration: "api", tree: [] }])).toBe("no dependencies declared\n");
});

File: togo/internal/cli/tree_test.go
package cli

// Port of src/cli/tree.test.ts.

import (
	"reflect"
	"strings"
	"testing"

	"github.com/nikeee/cappu/internal/packages"
)

func treeCoord(spec string) packages.Coordinates {
	p := strings.Split(spec, ":")
	return packages.NewCoordinates(p[0], p[1], p[2])
}

func treeSource() *packages.InMemoryPackageSource {
	pkg := func(spec string, deps ...string) packages.PackageMetadata {
		decls := make([]packages.DependencyDeclaration, 0, len(deps))
		for _, d := range deps {
			decls = append(decls, packages.DependencyDeclaration{Coordinates: treeCoord(d)})
		}
		return packages.PackageMetadata{Coordinates: treeCoord(spec), Dependencies: decls}
	}
	return packages.NewInMemoryPackageSource("registry", []packages.PackageMetadata{
		pkg("org.x:app:1.0", "org.x:lib:2.0", "org.y:util:3.0"),
		pkg("org.x:lib:2.0", "org.y:util:3.0"),
		pkg("org.y:util:3.0"),
	})
}

func TestBuildForestNestsTransitiveUnderRequester(t *testing.T) {
	res, err := packages.ResolveTransitive([]packages.Coordinates{treeCoord("org.x:app:1.0")},
		[]packages.PackageSource{treeSource()}, nil)
	if err != nil {
		t.Fatal(err)
	}
	// org.y:util:3.0 is reached first via org.x:app (nearest wins), so it nests
	// there and not again under org.x:lib.
	want := []treeNode{{
		Coordinate: "org.x:app:1.0",
		Dependencies: []treeNode{
			{Coordinate: "org.x:lib:2.0", Dependencies: []treeNode{}},
			{Coordinate: "org.y:util:3.0", Dependencies: []treeNode{}},
		},
	}}
	if got := buildForest(res); !reflect.DeepEqual(got, want) {
		t.Errorf("buildForest = %+v", got)
	}
}

func TestBuildForestMarksUnresolvedRoot(t *testing.T) {
	res, err := packages.ResolveTransitive([]packages.Coordinates{treeCoord("org.z:missing:9.9")},
		[]packages.PackageSource{treeSource()}, nil)
	if err != nil {
		t.Fatal(err)
	}
	want := []treeNode{{Coordinate: "org.z:missing:9.9", Dependencies: []treeNode{}, Unresolved: true}}
	if got := buildForest(res); !reflect.DeepEqual(got, want) {
		t.Errorf("buildForest = %+v", got)
	}
}

func TestFormatTreeRendersIndentedSections(t *testing.T) {
	sections := []treeSection{
		{Configuration: "api", Tree: []treeNode{{
			Coordinate: "org.x:app:1.0",
			Dependencies: []treeNode{
				{Coordinate: "org.x:lib:2.0", Dependencies: []treeNode{}},
				{Coordinate: "org.y:util:3.0", Dependencies: []treeNode{}, Unresolved: true},
			},
		}}},
		{Configuration: "implementation", Tree: []treeNode{}}, // empty: skipped
	}
	want := strings.Join([]string{
		"api",
		"└── org.x:app:1.0",
		"    ├── org.x:lib:2.0",
		"    └── org.y:util:3.0 (unresolved)",
		"",
	}, "\n")
	if got := FormatTree(sections, plainPaint); got != want {
		t.Errorf("FormatTree = %q", got)
	}
}

func TestFormatTreeNothingDeclared(t *testing.T) {
	got := FormatTree([]treeSection{{Configuration: "api", Tree: []treeNode{}}}, plainPaint)
	if got != "no dependencies declared\n" {
		t.Errorf("FormatTree = %q", got)
	}
}
